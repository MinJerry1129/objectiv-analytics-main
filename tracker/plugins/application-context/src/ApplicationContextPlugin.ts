/*
 * Copyright 2022 Objectiv B.V.
 */

import { ApplicationContext } from '@objectiv/schema';
import {
  ContextsConfig,
  makeApplicationContext,
  TrackerInterface,
  TrackerPluginInterface,
} from '@objectiv/tracker-core';

/**
 * The ApplicationContextPlugin adds an ApplicationContext as GlobalContext before events are transported.
 * ApplicationContext Validation is performed by OpenTaxonomyValidationPlugin
 */
export class ApplicationContextPlugin implements TrackerPluginInterface {
  readonly pluginName = `ApplicationContextPlugin`;
  initialized: boolean = false;
  applicationContext?: ApplicationContext;

  /**
   * Generates a ApplicationContext with the Tracker's applicationId.
   */
  initialize(tracker: TrackerInterface) {
    this.applicationContext = makeApplicationContext({
      id: tracker.applicationId,
    });

    this.initialized = true;

    if (globalThis.objectiv.devTools) {
      globalThis.objectiv.devTools.TrackerConsole.groupCollapsed(`｢objectiv:${this.pluginName}｣ Initialized`);
      globalThis.objectiv.devTools.TrackerConsole.log(`Application ID: ${tracker.applicationId}`);
      globalThis.objectiv.devTools.TrackerConsole.group(`Application Context:`);
      globalThis.objectiv.devTools.TrackerConsole.log(this.applicationContext);
      globalThis.objectiv.devTools.TrackerConsole.groupEnd();
      globalThis.objectiv.devTools.TrackerConsole.groupEnd();
    }
  }

  /**
   * Add the ApplicationContext to the Event's Global Contexts
   */
  enrich(contexts: Required<ContextsConfig>): void {
    if (!this.initialized) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `｢objectiv:${this.pluginName}｣ Cannot enrich. Make sure to initialize the plugin first.`
      );
    }
    if (this.applicationContext) {
      contexts.global_contexts.push(this.applicationContext);
    }
  }

  /**
   * Make this plugin always usable
   */
  isUsable(): boolean {
    return true;
  }
}
