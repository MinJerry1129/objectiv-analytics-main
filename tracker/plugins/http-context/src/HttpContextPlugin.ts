/*
 * Copyright 2021-2022 Objectiv B.V.
 */

import {
  GlobalContextName,
  makeHttpContext,
  TrackerEvent,
  TrackerInterface,
  TrackerPluginInterface,
  TrackerValidationRuleInterface,
} from '@objectiv/tracker-core';

/**
 * The HttpContext Plugin gathers the current URL using the Location API.
 * It implements the `initialize` lifecycle method. This ensures the Context is generated when the tracker is created.
 */
export class HttpContextPlugin implements TrackerPluginInterface {
  readonly pluginName = `HttpContextPlugin`;
  validationRules: TrackerValidationRuleInterface[] = [];
  initialized: boolean = false;

  /**
   * Generates an HttpContext and initializes validation rules.
   */
  initialize({ global_contexts, platform }: TrackerInterface): void {
    if (!this.isUsable()) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `｢objectiv:${
          this.pluginName
        }｣ Cannot initialize. Plugin is not usable (document: ${typeof document}, navigator: ${typeof navigator}).`
      );
      return;
    }

    if (globalThis.objectiv.devTools) {
      this.validationRules = [
        globalThis.objectiv.devTools.makeMissingGlobalContextValidationRule({
          platform,
          logPrefix: this.pluginName,
          contextName: GlobalContextName.HttpContext,
        }),
      ];
    }

    const httpContext = makeHttpContext({
      id: 'http_context',
      referrer: document.referrer ?? '',
      user_agent: navigator.userAgent ?? '',
    });

    global_contexts.push(httpContext);

    this.initialized = true;

    globalThis.objectiv.devTools?.TrackerConsole.log(
      `%c｢objectiv:${this.pluginName}｣ Initialized`,
      'font-weight: bold'
    );
  }

  /**
   * If the Plugin is usable runs all validation rules.
   */
  validate(event: TrackerEvent): void {
    if (!this.isUsable()) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `｢objectiv:${
          this.pluginName
        }｣ Cannot validate. Plugin is not usable (document: ${typeof document}, navigator: ${typeof navigator}).`
      );
      return;
    }

    if (!this.initialized) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `｢objectiv:${this.pluginName}｣ Cannot validate. Make sure to initialize the plugin first.`
      );
      return;
    }

    this.validationRules.forEach((validationRule) => validationRule.validate(event));
  }

  /**
   * Make this plugin usable only on web, eg: Document and Navigation APIs are both available
   */
  isUsable(): boolean {
    return typeof document !== 'undefined' && typeof navigator !== 'undefined';
  }
}
