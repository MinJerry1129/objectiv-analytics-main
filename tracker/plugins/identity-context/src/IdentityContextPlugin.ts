/*
 * Copyright 2022 Objectiv B.V.
 */

import { ContextsConfig, makeIdentityContext, TrackerPluginInterface } from '@objectiv/tracker-core';

/**
 * The required attributes of IdentityContext.
 */
export type IdentityContextAttributes = {
  id: string;
  value: string;
};

/**
 * The configuration object of IdentityContextPlugin.
 * Valid parameters are:
 *  - an IdentityContextAttributes object
 *  - an array of IdentityContextAttributes objects
 *  - a function returning an IdentityContextAttributes object
 *  - a function returning an array of IdentityContextAttributes objects
 */
export type IdentityContextPluginConfig =
  | IdentityContextAttributes
  | IdentityContextAttributes[]
  | (() => IdentityContextAttributes | IdentityContextAttributes[]);

/**
 * The IdentityContextPlugin adds one or more IdentityContexts as GlobalContexts before events are transported.
 */
export class IdentityContextPlugin implements TrackerPluginInterface {
  readonly pluginName = `IdentityContextPlugin`;
  readonly config: IdentityContextPluginConfig;

  /**
   * The constructor is merely responsible for processing the given configuration.
   */
  constructor(config: IdentityContextPluginConfig) {
    this.config = config;

    globalThis.objectiv.devTools?.TrackerConsole.log(
      `%c｢objectiv:${this.pluginName}｣ Initialized`,
      'font-weight: bold'
    );
  }

  /**
   * Factor and add the IdentityContext to the Event's Global Contexts
   */
  enrich(contexts: Required<ContextsConfig>): void {
    const identityContextAttributes = typeof this.config === 'function' ? this.config() : this.config;

    const identityContexts = (
      Array.isArray(identityContextAttributes) ? identityContextAttributes : [identityContextAttributes]
    ).map(makeIdentityContext);

    contexts.global_contexts.push(...identityContexts);
  }

  /**
   * Make this plugin always usable
   */
  isUsable(): boolean {
    return true;
  }
}
