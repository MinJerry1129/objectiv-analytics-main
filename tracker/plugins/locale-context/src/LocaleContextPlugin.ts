/*
 * Copyright 2022 Objectiv B.V.
 */

import { ContextsConfig, makeLocaleContext, TrackerPluginInterface } from '@objectiv/tracker-core';
import { CountryCodes } from './generated/CountryCodes';
import { LanguageCodes } from './generated/LanguageCodes';

/**
 * Factory functions may return string, null or undefined
 */
export type LocaleContextPluginFactoryFunction = () => string | null | undefined;

/**
 * The configuration object of LocaleContextPlugin. Developers must specify at least one of the factory functions.
 */
export type LocaleContextPluginConfig =
  | {
      idFactoryFunction: LocaleContextPluginFactoryFunction;
      languageFactoryFunction?: LocaleContextPluginFactoryFunction;
      countryFactoryFunction?: LocaleContextPluginFactoryFunction;
    }
  | {
      idFactoryFunction?: LocaleContextPluginFactoryFunction;
      languageFactoryFunction: LocaleContextPluginFactoryFunction;
      countryFactoryFunction?: LocaleContextPluginFactoryFunction;
    }
  | {
      idFactoryFunction?: LocaleContextPluginFactoryFunction;
      languageFactoryFunction?: LocaleContextPluginFactoryFunction;
      countryFactoryFunction: LocaleContextPluginFactoryFunction;
    };

/**
 * The LocaleContext Plugin executes the given idFactoryFunction to retrieve the identifier to factor a
 * LocaleContext which is attached to the Event's Global Contexts.
 * It implements the `enrich` lifecycle method. This ensures the locale is determined before each Event is sent.
 */
export class LocaleContextPlugin implements TrackerPluginInterface {
  readonly pluginName = `LocaleContextPlugin`;
  readonly initialized: boolean = false;
  readonly idFactoryFunction: LocaleContextPluginFactoryFunction = () => null;
  readonly languageFactoryFunction: LocaleContextPluginFactoryFunction = () => null;
  readonly countryFactoryFunction: LocaleContextPluginFactoryFunction = () => null;

  /**
   * The constructor is merely responsible for processing the given LocaleContextPluginConfig.
   */
  constructor(config: LocaleContextPluginConfig) {
    if (!config?.idFactoryFunction && !config?.countryFactoryFunction && !config?.languageFactoryFunction) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `%c｢objectiv:${this.pluginName}｣ Please specify at least one factory function.`,
        'font-weight: bold'
      );
      return;
    }

    this.idFactoryFunction = config.idFactoryFunction ?? this.idFactoryFunction;
    this.languageFactoryFunction = config.languageFactoryFunction ?? this.languageFactoryFunction;
    this.countryFactoryFunction = config.countryFactoryFunction ?? this.countryFactoryFunction;
    this.initialized = true;

    globalThis.objectiv.devTools?.TrackerConsole.log(
      `%c｢objectiv:${this.pluginName}｣ Initialized`,
      'font-weight: bold'
    );
  }

  /**
   * Generate a fresh LocaleContext before each TrackerEvent is handed over to the TrackerTransport.
   *
   */
  enrich(contexts: Required<ContextsConfig>): void {
    if (!this.initialized) {
      globalThis.objectiv.devTools?.TrackerConsole.warn(
        `%c｢objectiv:${this.pluginName}｣ Skipped enrich. Please initialize with at least one factory faction.`,
        'font-weight: bold'
      );
      return;
    }

    // Generate and validate language code.
    const maybeLanguageCode = this.languageFactoryFunction();
    const languageCode = maybeLanguageCode && LanguageCodes.includes(maybeLanguageCode) ? maybeLanguageCode : null;
    if (maybeLanguageCode && !languageCode) {
      globalThis.objectiv.devTools?.TrackerConsole.warn(
        `%c｢objectiv:${this.pluginName}｣ Language code is not ISO 639-1. Got: ${maybeLanguageCode}.`,
        'font-weight: bold'
      );
      return;
    }

    // Generate and validate country code.
    const maybeCountryCode = this.countryFactoryFunction();
    const countryCode = maybeCountryCode && CountryCodes.includes(maybeCountryCode) ? maybeCountryCode : null;
    if (maybeCountryCode && !countryCode) {
      globalThis.objectiv.devTools?.TrackerConsole.warn(
        `%c｢objectiv:${this.pluginName}｣ Country code is not ISO 3166-1 alpha-2. Got: ${maybeCountryCode}.`,
        'font-weight: bold'
      );
      return;
    }

    // Generate LocaleContext id. We use either what the developer specified, or make one with language and country.
    let localeContextId = this.idFactoryFunction();
    if (!localeContextId) {
      if (languageCode && countryCode) {
        localeContextId = `${languageCode}_${countryCode}`;
      } else {
        localeContextId = `${languageCode ?? ''}${countryCode ?? ''}`;
      }
    }

    if (!localeContextId) {
      globalThis.objectiv.devTools?.TrackerConsole.warn(
        `%c｢objectiv:${this.pluginName}｣ Cannot enrich. Could not generate a valid LocaleContext.`,
        'font-weight: bold'
      );
      return;
    }

    const localeContext = makeLocaleContext({
      id: localeContextId,
      language_code: languageCode,
      country_code: countryCode,
    });
    contexts.global_contexts.push(localeContext);
  }

  /**
   * Make this plugin always usable. The provided idFactoryFunction may return `null` when detection is impossible.
   */
  isUsable(): boolean {
    return true;
  }
}
