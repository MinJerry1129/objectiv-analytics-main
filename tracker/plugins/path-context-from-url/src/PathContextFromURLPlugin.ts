/*
 * Copyright 2021-2022 Objectiv B.V.
 */

import { ContextsConfig, makePathContext, TrackerPluginInterface } from '@objectiv/tracker-core';

/**
 * The PathContextFromURL Plugin gathers the current URL using the Location API.
 * It implements the `enrich` lifecycle method. This ensures the URL is retrieved before each Event is sent.
 *
 * Event Validation:
 *  - Must be present in Global Contexts
 *  - Must not be present multiple times
 */
export class PathContextFromURLPlugin implements TrackerPluginInterface {
  readonly pluginName = `PathContextFromURLPlugin`;

  /**
   * Initializes validation rules.
   */
  initialize(): void {
    globalThis.objectiv.devTools?.TrackerConsole.log(
      `%c｢objectiv:${this.pluginName}｣ Initialized`,
      'font-weight: bold'
    );
  }

  /**
   * Generate a fresh PathContext before each TrackerEvent is handed over to the TrackerTransport.
   */
  enrich(contexts: Required<ContextsConfig>): void {
    if (!this.isUsable()) {
      globalThis.objectiv.devTools?.TrackerConsole.error(
        `｢objectiv:${this.pluginName}｣ Cannot enrich. Plugin is not usable (document: ${typeof document}).`
      );
      return;
    }

    const pathContext = makePathContext({
      id: document.location.href,
    });
    contexts.global_contexts.push(pathContext);
  }

  /**
   * Make this plugin usable only on web, eg: Document and Location APIs are both available
   */
  isUsable(): boolean {
    return typeof document !== 'undefined' && typeof document.location !== 'undefined';
  }
}
