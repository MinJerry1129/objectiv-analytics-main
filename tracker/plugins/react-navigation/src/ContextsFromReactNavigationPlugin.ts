/*
 * Copyright 2022 Objectiv B.V.
 */

import { makeRootLocationContext, TrackerPluginInterface } from '@objectiv/tracker-core';
import { ContextsConfig, makePathContext } from '@objectiv/tracker-core';
import { getPathFromState, NavigationContainerRefWithCurrent } from '@react-navigation/native';

/**
 * The configuration of ContextsFromReactNavigationPlugin always requires a React NavigationContainerRef.
 */
export type ContextsFromReactNavigationPluginConfig<ParamList extends ReactNavigation.RootParamList> = {
  navigationContainerRef: NavigationContainerRefWithCurrent<ParamList>;
};

/**
 * A plugin that can infer RootLocation and Path Contexts by querying React Navigation NavigationContainerRef state.
 */
export class ContextsFromReactNavigationPlugin<ParamList extends ReactNavigation.RootParamList>
  implements TrackerPluginInterface
{
  readonly pluginName = `ContextsFromReactNavigationPlugin`;
  readonly navigationContainerRef: NavigationContainerRefWithCurrent<ParamList>;

  /**
   * The constructor is responsible for processing the given configuration.
   */
  constructor(config: ContextsFromReactNavigationPluginConfig<ParamList>) {
    this.navigationContainerRef = config.navigationContainerRef;

    globalThis.objectiv.devTools?.TrackerConsole.log(
      `%c｢objectiv:${this.pluginName}｣ Initialized`,
      'font-weight: bold'
    );
  }

  /**
   * Generate RootLocationContext and PathContext from React Navigation Container ref state.
   */
  enrich(contexts: Required<ContextsConfig>) {
    let rootLocationContextId;
    let pathContextId;

    if (this.navigationContainerRef.isReady()) {
      const currentRoute = this.navigationContainerRef.getCurrentRoute();
      if (currentRoute && currentRoute.name) {
        rootLocationContextId = currentRoute.name;
      }
      pathContextId = getPathFromState(this.navigationContainerRef.getRootState());
    }

    contexts.location_stack.unshift(makeRootLocationContext({ id: rootLocationContextId ?? 'home' }));
    contexts.global_contexts.push(makePathContext({ id: pathContextId ?? '/' }));
  }

  /**
   * This plugin is always usable. If React Navigation is not ready it will still produce valid Contexts.
   */
  isUsable(): boolean {
    return true;
  }
}
