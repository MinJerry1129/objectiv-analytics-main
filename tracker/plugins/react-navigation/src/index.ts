/*
 * Copyright 2022 Objectiv B.V.
 */

export * from './ContextsFromReactNavigationPlugin';
export * from './makeLinkContextProps';
export * from './makeLinkPressListener';
export * from './TrackedLink';
