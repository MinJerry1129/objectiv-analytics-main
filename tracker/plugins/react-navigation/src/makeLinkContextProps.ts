/*
 * Copyright 2022 Objectiv B.V.
 */

import { makeId } from '@objectiv/tracker-core';
import { makeTitleFromChildren } from '@objectiv/tracker-react-core';
import { getPathFromState } from '@react-navigation/native';
import { TrackedLinkProps } from './TrackedLink';

/**
 * A helper method to generate LinkContext `id` and `href` props from TrackedLink `to` and `children` props.
 */
export function makeLinkContextProps<ParamList extends ReactNavigation.RootParamList>(
  props: TrackedLinkProps<ParamList>
) {
  // Either use the given id or attempt to auto-detect `id` for LinkContext by looking at the `children` prop.
  const title = makeTitleFromChildren(props.children);
  const contextId = props.id ?? makeId(title);

  // Use React Navigation `getPathFromState` to generate the `href` prop. Unless it was already a string.
  let contextHref: string;
  if (typeof props.to === 'string') {
    contextHref = props.to as string;
  } else {
    contextHref = getPathFromState({
      routes: [
        {
          name: props.to.screen,
          params: props.to.params as {},
        },
      ],
    });
  }

  return {
    contextId,
    contextHref,
  };
}
