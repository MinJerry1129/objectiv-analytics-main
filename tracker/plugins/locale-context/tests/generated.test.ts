/*
 * Copyright 2022 Objectiv B.V.
 */

import { CountryCodes, LanguageCodes } from '../src';

describe('Generated', () => {
  it('LanguageCodes should match snapshot', async () => {
    expect(LanguageCodes).toMatchSnapshot();
  });
  it('CountryCodes should match snapshot', async () => {
    expect(CountryCodes).toMatchSnapshot();
  });
});
