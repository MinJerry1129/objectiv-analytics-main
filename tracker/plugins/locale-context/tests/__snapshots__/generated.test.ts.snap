// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated CountryCodes should match snapshot 1`] = `
Array [
  "AD",
  "AE",
  "AF",
  "AG",
  "AL",
  "AM",
  "AO",
  "AQ",
  "AR",
  "AS",
  "AT",
  "AU",
  "AW",
  "AX",
  "AZ",
  "BA",
  "BB",
  "BD",
  "BE",
  "BF",
  "BG",
  "BH",
  "BI",
  "BJ",
  "BL",
  "BM",
  "BN",
  "BO",
  "BR",
  "BS",
  "BT",
  "BV",
  "BW",
  "BY",
  "BZ",
  "CA",
  "CC",
  "CD",
  "CF",
  "CG",
  "CH",
  "CI",
  "CK",
  "CL",
  "CM",
  "CN",
  "CO",
  "CR",
  "CU",
  "CV",
  "CW",
  "CX",
  "CY",
  "CZ",
  "DE",
  "DJ",
  "DK",
  "DM",
  "DO",
  "DZ",
  "EC",
  "EE",
  "EG",
  "EH",
  "ER",
  "ES",
  "ET",
  "FI",
  "FJ",
  "FK",
  "FM",
  "FO",
  "FR",
  "GA",
  "GB",
  "GD",
  "GF",
  "GG",
  "GH",
  "GI",
  "GL",
  "GM",
  "GN",
  "GP",
  "GQ",
  "GR",
  "GS",
  "GT",
  "GU",
  "GW",
  "GY",
  "HK",
  "HM",
  "HN",
  "HR",
  "HT",
  "HU",
  "ID",
  "IE",
  "IL",
  "IM",
  "IN",
  "IO",
  "IQ",
  "IR",
  "IS",
  "IT",
  "JE",
  "JM",
  "JO",
  "JP",
  "KE",
  "KG",
  "KH",
  "KI",
  "KM",
  "KN",
  "KP",
  "KR",
  "KW",
  "KY",
  "KZ",
  "LA",
  "LB",
  "LC",
  "LI",
  "LK",
  "LR",
  "LS",
  "LU",
  "LV",
  "LY",
  "MA",
  "MC",
  "MD",
  "MF",
  "MG",
  "MH",
  "MK",
  "ML",
  "MM",
  "MN",
  "MO",
  "MP",
  "MQ",
  "MR",
  "MS",
  "MT",
  "MU",
  "MV",
  "MW",
  "MX",
  "MY",
  "MZ",
  "NA",
  "NC",
  "NE",
  "NF",
  "NG",
  "NI",
  "NL",
  "NO",
  "NP",
  "NR",
  "NU",
  "NZ",
  "OM",
  "PA",
  "PE",
  "PF",
  "PG",
  "PH",
  "PK",
  "PL",
  "PM",
  "PN",
  "PR",
  "PS",
  "PT",
  "PW",
  "PY",
  "QA",
  "RE",
  "RO",
  "RS",
  "RW",
  "SA",
  "SB",
  "SC",
  "SD",
  "SE",
  "SG",
  "SH",
  "SI",
  "SJ",
  "SL",
  "SM",
  "SN",
  "SO",
  "SR",
  "SS",
  "ST",
  "SV",
  "SX",
  "SY",
  "SZ",
  "TC",
  "TD",
  "TF",
  "TG",
  "TH",
  "TJ",
  "TK",
  "TL",
  "TM",
  "TN",
  "TO",
  "TR",
  "TT",
  "TV",
  "TW",
  "TZ",
  "UA",
  "UG",
  "UM",
  "US",
  "UY",
  "UZ",
  "VA",
  "VC",
  "VE",
  "VG",
  "VI",
  "VN",
  "VU",
  "WF",
  "WS",
  "YE",
  "YT",
  "ZA",
  "ZM",
  "ZW",
]
`;

exports[`Generated LanguageCodes should match snapshot 1`] = `
Array [
  "aa",
  "ab",
  "ae",
  "af",
  "ak",
  "am",
  "an",
  "ar",
  "as",
  "av",
  "ay",
  "az",
  "ba",
  "be",
  "bg",
  "bh",
  "bi",
  "bm",
  "bn",
  "bo",
  "br",
  "bs",
  "ca",
  "ce",
  "ch",
  "co",
  "cr",
  "cs",
  "cu",
  "cv",
  "cy",
  "da",
  "de",
  "dv",
  "dz",
  "ee",
  "el",
  "en",
  "eo",
  "es",
  "et",
  "eu",
  "fa",
  "ff",
  "fi",
  "fj",
  "fo",
  "fr",
  "fy",
  "ga",
  "gd",
  "gl",
  "gn",
  "gu",
  "gv",
  "ha",
  "he",
  "hi",
  "ho",
  "hr",
  "ht",
  "hu",
  "hy",
  "hz",
  "ia",
  "id",
  "ie",
  "ig",
  "ii",
  "ik",
  "io",
  "is",
  "it",
  "iu",
  "ja",
  "jv",
  "ka",
  "kg",
  "ki",
  "kj",
  "kk",
  "kl",
  "km",
  "kn",
  "ko",
  "kr",
  "ks",
  "ku",
  "kv",
  "kw",
  "ky",
  "la",
  "lb",
  "lg",
  "li",
  "ln",
  "lo",
  "lt",
  "lu",
  "lv",
  "mg",
  "mh",
  "mi",
  "mk",
  "ml",
  "mn",
  "mr",
  "ms",
  "mt",
  "my",
  "na",
  "nb",
  "nd",
  "ne",
  "ng",
  "nl",
  "nn",
  "no",
  "nr",
  "nv",
  "ny",
  "oc",
  "oj",
  "om",
  "or",
  "os",
  "pa",
  "pi",
  "pl",
  "ps",
  "pt",
  "qu",
  "rm",
  "rn",
  "ro",
  "ru",
  "rw",
  "sa",
  "sc",
  "sd",
  "se",
  "sg",
  "si",
  "sk",
  "sl",
  "sm",
  "sn",
  "so",
  "sq",
  "sr",
  "ss",
  "st",
  "su",
  "sv",
  "sw",
  "ta",
  "te",
  "tg",
  "th",
  "ti",
  "tk",
  "tl",
  "tn",
  "to",
  "tr",
  "ts",
  "tt",
  "tw",
  "ty",
  "ug",
  "uk",
  "ur",
  "uz",
  "ve",
  "vi",
  "vo",
  "wa",
  "wo",
  "xh",
  "yi",
  "yo",
  "za",
  "zh",
  "zu",
]
`;
